### PR TITLE
Additions to "Implement better support for MIR"

### PR DESCRIPTION
**Pull request by W95Psp**

### CODE REVIEW

File: frontend/exporter/src/types/copied.rs [hunk 1]
```diff
@@ -353,55 +205,38 @@ pub struct Scope {
...
-                let kind = ConstantExprKind::GlobalName { id };
-                Decorated {
-                    ty: ty.sinto(s),
-                    span: span.sinto(s),
-                    contents: Box::new(kind),
-                    hir_id: None,
-                    attributes: vec![],
+                if is_anon_const(ucv.def, tcx) {
+                    let evaluated_c = self.eval(tcx, get_param_env(s));
+                    if &evaluated_c == self {
```
Comment (sonmarcho) [inline frontend/exporter/src/types/copied.rs:223]: Are you sure this works? Shouldn't we do `if let ...::Unevaluated = evaluated_c { ... }`?

File: frontend/exporter/src/constant_utils.rs [hunk 2]
```diff
@@ -204,76 +192,72 @@ pub(crate) fn scalar_to_constant_expr<'tcx, S: BaseState<'tcx>>(
...
-        ty: ty.sinto(s),
-        span: cspan,
-        contents: Box::new(kind),
-        hir_id: None,
-        attributes: vec![],
-    }
+    kind.decorate(ty.sinto(s), cspan)
 }
 
-/// We evaluate the anonymous constants
```
Comment (sonmarcho) [inline frontend/exporter/src/constant_utils.rs:236]: The reason why I had put written this function was to make sure the evaluation policy was consistent across the code base.

Comment (W95Psp) [reply frontend/exporter/src/constant_utils.rs:236]: Makes sense indeed! Just factored out the handling of unevaluated constants in 0514b83b4bd2965e9569a745919eb3a9b492999b

File: frontend/exporter/src/types/copied.rs (hunk 1 above)
Comment (W95Psp) [reply frontend/exporter/src/types/copied.rs]: Yes, the implementation of the `eval` methods returns the constant as it is

File: frontend/exporter/src/types/copied.rs [hunk 3]
```diff
@@ -1791,6 +1581,7 @@ pub struct VariantInformations {
 
     pub typ: DefId,
     pub variant: DefId,
+    pub variant_index: VariantIdx,
```
Comment (sonmarcho) [inline frontend/exporter/src/types/copied.rs:1584]: I need this to be an option : `None` if the ADT is a structure

**Review by sonmarcho — APPROVED** [review]
Looks good! You can merge whenever you want, modulo the comment I left for `VariantInformation` :)

File: frontend/exporter/src/types/copied.rs (hunk 3 above)
Comment (W95Psp) [reply frontend/exporter/src/types/copied.rs:1584]: I removed that option because of the field `typ_is_struct` below: isn't that what you need?

Comment (sonmarcho) [reply frontend/exporter/src/types/copied.rs:1584]: Indeed, I hadn't noticed. That works for me.

**Pull request merged**